tools/blocksconvert/builder/tsdb.go

```diff
@@ -223,16 +243,39 @@ func (d *tsdbBuilder) finishBlock(source string, labels map[string]string) (ulid
 
 	level.Info(d.log).Log("msg", "added symbols to index", "count", symbols)
 
-	stats, err := addSeriesToIndex(indexWriter, d.series)
+	unsortedChunksReader, err := chunks.NewDirReader(filepath.Join(d.tmpBlockDir, unsortedChunksDir), nil)
+	if err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "chunks reader")
+	}
+	toClose[unsortedChunksReaderName] = unsortedChunksReader
+
+	chunksWriter, err := chunks.NewWriter(filepath.Join(d.tmpBlockDir, "chunks"))
+	if err != nil {
+		return ulid.ULID{}, errors.Wrap(err, "chunks reader")
+	}
+	toClose[chunksWriterName] = chunksWriter
+
+	stats, err := addSeriesToIndex(indexWriter, d.series, unsortedChunksReader, chunksWriter)
 	if err != nil {
 		return ulid.ULID{}, errors.Wrap(err, "adding series")
 	}
 	meta.Stats = stats
 
 	level.Info(d.log).Log("msg", "added series to index", "series", stats.NumSeries, "chunks", stats.NumChunks, "samples", stats.NumSamples)
 
-	if err := indexWriter.Close(); err != nil {
-		return ulid.ULID{}, errors.Wrap(err, "closing index writer")
+	// Close index writer, unsorted chunks reader and chunks writer.
+	for k, c := range toClose {
+		delete(toClose, k)
+
+		err := c.Close()
+		if err != nil {
+			return ulid.ULID{}, errors.Wrapf(err, "closing %s", k)
+		}
+	}
+
+	// Delete unsorted chunks, they are no longer needed.
+	if err := os.RemoveAll(filepath.Join(d.tmpBlockDir, unsortedChunksDir)); err != nil {
```

Comment:
Should be done in the `defer` cleanup function too? The one defined above.

Reply:
Defer only closes open resources on error, but doesn't modify the outcome.

Deleting unsorted chunks is part of normal flow, which can end up with error, although it will just make metrics wrong (when counting block dir). Still, I'd keep it as-is.